### PR TITLE
CSHARP-4486: Add support for bitwise operators.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/AstNodeType.cs [hunk 1]
```diff
@@ -20,7 +20,6 @@ internal enum AstNodeType
         AccumulatorField,
         AddFieldsStage,
         AllFilterOperation,
-        AndExpression,
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/AstNodeType.cs:23]: $and and $or are just nary operators, we don't really need special AST classes for them.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Expressions/AstExpression.cs [hunk 2]
```diff
@@ -80,24 +80,8 @@ public static AstExpression Add(params AstExpression[] args)
...
-                    else
-                    {
-                        flattenedArgs.Add(arg);
-                    }
-                }
-                return new AstNaryExpression(AstNaryOperator.Add, flattenedArgs);
-            }
-
-            return new AstNaryExpression(AstNaryOperator.Add, args);
+            var flattenedArgs = FlattenNaryArgs(args, AstNaryOperator.Add);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Expressions/AstExpression.cs:83]: There was a lot of code duplication related to flattening nary operators which could be centralized in a new helper method.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Expressions/AstNaryOperator.cs [hunk 3]
```diff
@@ -20,13 +20,18 @@ namespace MongoDB.Driver.Linq.Linq3Implementation.Ast.Expressions
     internal enum AstNaryOperator
     {
         Add,
+        And,
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Expressions/AstNaryOperator.cs:23]: $and and $or are just nary operators now, not modeled with separate AST classes.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/BinaryExpressionToAggregationExpressionTranslator.cs [hunk 4]
```diff
@@ -81,19 +81,24 @@ public static AggregationExpression Translate(TranslationContext context, Binary
                 ExpressionType.Add => IsStringConcatenationExpression(expression) ?
                     AstExpression.Concat(leftTranslation.Ast, rightTranslation.Ast) :
                     AstExpression.Add(leftTranslation.Ast, rightTranslation.Ast),
-                ExpressionType.And => AstExpression.And(leftTranslation.Ast, rightTranslation.Ast),
+                ExpressionType.And => expression.Type == typeof(bool) ?
+                    AstExpression.And(leftTranslation.Ast, rightTranslation.Ast) :
+                    AstExpression.BitAnd(leftTranslation.Ast, rightTranslation.Ast),
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/BinaryExpressionToAggregationExpressionTranslator.cs:86]: C# uses `Expression.And` for both boolean and bit-wise operations, but MQL uses different operators for each.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/BinaryExpressionToAggregationExpressionTranslator.cs [hunk 5]
```diff
@@ -81,19 +81,24 @@ public static AggregationExpression Translate(TranslationContext context, Binary
...
+                ExpressionType.ExclusiveOr => AstExpression.BitXor(leftTranslation.Ast, rightTranslation.Ast),
                 ExpressionType.GreaterThan => AstExpression.Gt(leftTranslation.Ast, rightTranslation.Ast),
                 ExpressionType.GreaterThanOrEqual => AstExpression.Gte(leftTranslation.Ast, rightTranslation.Ast),
                 ExpressionType.LessThan => AstExpression.Lt(leftTranslation.Ast, rightTranslation.Ast),
                 ExpressionType.LessThanOrEqual => AstExpression.Lte(leftTranslation.Ast, rightTranslation.Ast),
                 ExpressionType.Modulo => AstExpression.Mod(leftTranslation.Ast, rightTranslation.Ast),
                 ExpressionType.Multiply => AstExpression.Multiply(leftTranslation.Ast, rightTranslation.Ast),
                 ExpressionType.NotEqual => AstExpression.Ne(leftTranslation.Ast, rightTranslation.Ast),
-                ExpressionType.Or => AstExpression.Or(leftTranslation.Ast, rightTranslation.Ast),
+                ExpressionType.Or => expression.Type == typeof(bool) ?
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/BinaryExpressionToAggregationExpressionTranslator.cs:99]: Same as above.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NotExpressionToAggregationExpressionTranslator.cs [hunk 6]
```diff
@@ -25,7 +25,7 @@ public static AggregationExpression Translate(TranslationContext context, UnaryE
             if (expression.NodeType == ExpressionType.Not)
             {
                 var operandTranslation = ExpressionToAggregationExpressionTranslator.Translate(context, expression.Operand);
-                var ast = AstExpression.Not(operandTranslation.Ast);
+                var ast = expression.Type == typeof(bool) ? AstExpression.Not(operandTranslation.Ast) : AstExpression.BitNot(operandTranslation.Ast);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/NotExpressionToAggregationExpressionTranslator.cs:28]: Same as above.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Linq.Linq3Implementation.Jira
+{
+    public class CSharp4486Tests : Linq3IntegrationTest
+    {
+        [Theory]
+        [ParameterAttributeData]
+        public void BitAnd_with_two_arguments_should_work(
+            [Values(LinqProvider.V2, LinqProvider.V3)] LinqProvider linqProvider)
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs:32]: Should we have some tests re using of new operators as filter expressions?

Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs:32]: Should we have some tests to validate if boolean operations are working as expected? As I've seen it is going though the same machinery and we might want to check if mql is being generated properly.

**Review by sanych-sun — APPROVED** [review]
LGTM + minor comments

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs (hunk 7 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs:32]: These operators are not supported in filters.

If these expressions were used in a `Where` they would just become `$expr` which would be testing the same thing again.

Given that, do you still think we should add `Where` tests?

Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs:32]: What do you mean?

These test *are* testing boolean operations, including the `result` returned.

Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs:32]: Oh I think I see what you mean. You are distinguishing between *bitwise* operators (with numeric arguments) and *boolean* operators (with boolean arguments).

Sure, I can add tests for those here.

**Review by sanych-sun — APPROVED** [review]
LGTM

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs (hunk 7 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4486Tests.cs:174]: As it was discussed offline - no needs to implement tests for Where expressions.

**Pull request merged**